### PR TITLE
Add cyclical learning rate

### PR DESCRIPTION
**Pull request by Tauranis**
Add cyclical learning rate as reported by `Leslie N Smith. Cyclical learning rates for training neural networks. arXiv preprint arXiv:1506.01186v3, 2016.`

See [this example](https://postimg.cc/image/vzwtaxzhj/) to understand what a cyclical learning rate is about

The function can be easily coupled to others like `tf.train.exponential`

  ```python
  ...
  cosine_lr = tf.train.cosine_cyclical(starter_learning_rate, global_step,
                                            decay_steps, end_learning_rate, T)
  cosine_exp_lr = tf.train.exponential_decay(cosine_lr,global_step,300,0.7,staircase=False)

  ```

  For the a better comprehension, see [this example](https://postimg.cc/image/8tcgv2xjb/)  of how the snippet above looks like in practice.

### CODE REVIEW

**Comment by qlzh727** [comment]
@wicke, didn't find a good reviewer for this change based on the change history, do u mind reroute this if you know anyone worked on this?

**Comment by Tauranis** [comment]
I'm happy that another contributor proposed another cyclical learning rate on PR #20785 based on the same reference. I just want to make it clear that the cyclical learning rate I've coded is based on a cosine curve, not linear as proposed by #20785

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @martinwicke: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 14 days with no activity and the `awaiting review` label has been applied.

**Comment by qlzh727** [comment]
Ping @martinwicke for review.

**Comment by futurely** [comment]
Have you compared the experiment results with cosine_decay_restarts and https://github.com/tensorflow/tensorflow/pull/20785?

https://github.com/tensorflow/tensorflow/blob/483913a32ef0807791a9b5edfb9628cd66c786a2/tensorflow/python/training/learning_rate_decay.py#L710

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @martinwicke: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 14 days with no activity and the `awaiting review` label has been applied.

**Comment by Tauranis** [comment]
Nope.

But as I mentioned before, the cyclical learning rate I've coded is based on a cosine curve, not linear as proposed by #20785

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @martinwicke: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 14 days with no activity and the `awaiting review` label has been applied.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @martinwicke: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 29 days with no activity and the `awaiting review` label has been applied.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @martinwicke: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 45 days with no activity and the `awaiting review` label has been applied.

**Comment by loshchil** [comment]
The image linked in the opening post shows SGDR and it is reproducible with cosine_decay_restarts (https://github.com/tensorflow/tensorflow/pull/11749).

**Comment by Tauranis** [comment]
> The image linked in the opening post shows SGDR and it is reproducible with cosine_decay_restarts (#11749).

I agree. There is no need to duplicate a function. I haven't seen this before. My bad =P . I'm gonna close this PR.
